File: relay-frontend/types/template.ts
export interface Question {
  id: string
  question: string
  type: 'open_ended' | 'yes_no' | 'multiple_choice'
  required: boolean
  follow_up?: Record<string, unknown>
}

export interface Template {
  id: string
  user_id: string
  name: string
  description: string | null
  voice_id: string
  initial_message: string
  questions: Question[]
  closing_message: string | null
  extraction_schema: Record<string, unknown>
  max_duration_seconds: number
  created_at: string
  updated_at: string
}

export interface CreateTemplateInput {
  name: string
  description?: string
  voice_id: string
  initial_message: string
  questions?: Question[]
  closing_message?: string
  extraction_schema?: Record<string, unknown>
  max_duration_seconds?: number
}

export interface UpdateTemplateInput {
  name?: string
  description?: string
  voice_id?: string
  initial_message?: string
  questions?: Question[]
  closing_message?: string
  extraction_schema?: Record<string, unknown>
  max_duration_seconds?: number
}

File: relay-backend/app/schemas/template.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from uuid import UUID

class QuestionSchema(BaseModel):
    id: str
    question: str
    type: str = "open_ended" # open_ended, yes_no, multiple_choice
    required: bool = True
    follow_up: Optional[dict] = None

class TemplateCreate(BaseModel):
    name: str
    description: Optional[str] = None
    voice_id: str
    initial_message: str
    questions: list[QuestionSchema] = Field(default_factory=list)
    closing_message: Optional[str]
    extraction_schema: dict = Field(default_factory=dict)
    max_duration_seconds: int = 300

class TemplateUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    voice_id: Optional[str] = None
    initial_message: Optional[str] = None
    questions: list[QuestionSchema] = None
    closing_message: Optional[str] = None
    extraction_schema: Optional[dict] = None
    max_duration_seconds: Optional[int] = None   

class TemplateResponse(BaseModel):
    id: UUID
    user_id: UUID
    name: str
    description: Optional[str]
    voice_id: str
    initial_message: str
    questions: list[dict]
    closing_message: Optional[str]
    extraction_schema: dict
    max_duration_seconds: int
    created_at: datetime
    updated_at: datetime